### PR TITLE
VORTEX-5544 Icon Builder

### PR DESCRIPTION
**Pull request by jkae**
## Fixes
- Move font initialization to enums.
- Misc. formatting updates.
- Complete `javafx.scene.paint.Color` to `java.awt.Color` conversion function.
- Add methods in font icon class to manually set x- and y-offset. Retains original functionality if not given.

## Proposed Changes
- Rename `FxAwesomeIcons` to `FxIcons`.
  - Additional CSS classes based on font name to allow use with any icon set.
- Replace `FontAwesomeIcon` etc. with generic icon class that can build from any icon enum.
  - Removes `AbstractFontIcon`.
- Remove `FontIcons` loader class.
- Add icon builder for creating icons from icon fonts.
screenshot from 2018-04-02 09-37-18

### CODE REVIEW

File: open-sphere-base/core/src/main/resources/styles/opensphere.css [hunk 1]
```diff
@@ -27,6 +27,18 @@
     -fx-text-fill: linear-gradient(to bottom, derive(-opensphere-dark, 20%) 10%, derive(-opensphere-dark, -40%) 80%);
 }
 
+.icons-FontAwesome {
+    -fx-font-family: FontAwesome !important;
```
Comment (brelsftm) [inline open-sphere-base/core/src/main/resources/styles/opensphere.css]: What rule is !important supposed to be overriding? This css seems pretty specific to each class and shouldnt need the !important tag

File: open-sphere-base/mantle/src/main/java/io/opensphere/mantle/icon/impl/gui/IconBuilderPane.java [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+    {
+        myOwner = owner;
+
+        setTop(createTop());
+
+        setCenter(createImageView());
+        BorderPane.setAlignment(myImageRenderView, Pos.CENTER);
+        myImageRenderView.autosize();
+
+//        myComboBox.setOnAction((event) -> updateImageView());
```
Comment (brelsftm) [inline open-sphere-base/mantle/src/main/java/io/opensphere/mantle/icon/impl/gui/IconBuilderPane.java]: Do you need this comment?

File: open-sphere-base/mantle/src/main/java/io/opensphere/mantle/icon/impl/gui/IconRotationPane.java [hunk 3]
```diff
@@ -11,9 +12,8 @@
 import javafx.scene.layout.HBox;
 import javafx.scene.layout.Priority;
 
-import io.opensphere.mantle.icon.IconRecord;
-
 /** Icon rotation pane. */
+@SuppressWarnings("restriction")
```
Comment (brelsftm) [inline open-sphere-base/mantle/src/main/java/io/opensphere/mantle/icon/impl/gui/IconRotationPane.java]: What is this doing ?

Comment (jkae) [reply open-sphere-base/mantle/src/main/java/io/opensphere/mantle/icon/impl/gui/IconRotationPane.java]: API restriction warnings. You're not "allowed" to use JavaFX unless you're within a JFX project, which this is technically not.

Seeing the warnings is obnoxious (look into anything in `io.opensphere.core.util.fx` as an example).

File: open-sphere-base/core/src/main/resources/styles/opensphere.css (hunk 1 above)
Comment (jkae) [reply open-sphere-base/core/src/main/resources/styles/opensphere.css]: They're applied to the same items as the `.icons` class.

File: open-sphere-base/core/src/main/java/io/opensphere/core/util/GovIcon.java [hunk 4]
```diff
@@ -433,10 +443,17 @@ private GovIcon(String pFontCode)
      * Gets the value of the {@link #myFontCode} field.
      *
      * @return the value stored in the {@link #myFontCode} field.
+     * @override
```
Comment (rkausch) [inline open-sphere-base/core/src/main/java/io/opensphere/core/util/GovIcon.java]: Not necessary

File: open-sphere-base/core/src/main/java/io/opensphere/core/util/MilitaryRankIcon.java [hunk 5]
```diff
@@ -238,9 +248,17 @@ private MilitaryRankIcon(String pFontCode)
      * Gets the value of the {@link #myFontCode} field.
      *
      * @return the value stored in the {@link #myFontCode} field.
+     * @override
```
Comment (rkausch) [inline open-sphere-base/core/src/main/java/io/opensphere/core/util/MilitaryRankIcon.java]: not necessary

File: open-sphere-base/core/src/main/java/io/opensphere/core/util/fx/FXUtilities.java [hunk 6]
```diff
@@ -453,8 +454,15 @@ public static Color fromAwtColor(java.awt.Color awtColor)
      */
     public static java.awt.Color toAwtColor(Color color)
     {
-        // TODO add opacity
-        return new java.awt.Color(color.hashCode() >> 8);
+        java.awt.Color awtColor = null;
+
+        if (color != null)
```
Comment (rkausch) [inline open-sphere-base/core/src/main/java/io/opensphere/core/util/fx/FXUtilities.java]: what do we do if the supplied color is null? Default to white? Can that be explicit?

File: open-sphere-base/core/src/main/java/io/opensphere/core/util/swing/GenericFontIcon.java [hunk 7]
```diff
@@ -28,7 +28,7 @@
     private static final int DEFAULT_SIZE = 16;
 
     /** A lock used to enforce synchronization. */
-    private static final Object PAINT_LOCK = new Object[0];
+    private final Object PAINT_LOCK = new Object[0];
```
Comment (rkausch) [inline open-sphere-base/core/src/main/java/io/opensphere/core/util/swing/GenericFontIcon.java]: Doesn't removing the static nature of the lock object allow for the creation of a lock per instance of the class, which would be contrary to what we want for concurrency protection?

File: open-sphere-base/mantle/src/main/java/io/opensphere/mantle/icon/impl/gui/IconBuilderChoiceDialog.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+     * Constructs the tree of icon types.
+     *
+     * @return the type tree
+     */
+    private JTree createInternalTree()
+    {
+        Vector<String> treeValues = new Vector<>();
+        // Temporarily hard-code icon values.
+        treeValues.add("FontAwesome Icons");
+        treeValues.add("Govicon Icons");
```
Comment (rkausch) [inline open-sphere-base/mantle/src/main/java/io/opensphere/mantle/icon/impl/gui/IconBuilderChoiceDialog.java]: "Govicon Icons" sounds a bit redundant. Can we call them "Government Icons"?

File: open-sphere-base/mantle/src/main/java/io/opensphere/mantle/icon/impl/gui/IconRotationPane.java (hunk 3 above)
Comment (rkausch) [reply open-sphere-base/mantle/src/main/java/io/opensphere/mantle/icon/impl/gui/IconRotationPane.java]: You can get around this by using the e(fx)clipse plugins

File: open-sphere-base/core/src/main/java/io/opensphere/core/util/swing/GenericFontIcon.java (hunk 7 above)
Comment (jkae) [reply open-sphere-base/core/src/main/java/io/opensphere/core/util/swing/GenericFontIcon.java]: The image rendering is per-instance of the GenericFontIcon class, so I think a static lock enforces unnecessary synchronization when rendering multiple icons. It doesn't make sense to me, but then I also am not entirely sure if Java uses the same Graphics2D space for every image. I think if it does, it shouldn't when using `createGraphics()`.

**Pull request merged**